subsys/bluetooth/host/Kconfig

```diff
@@ -129,6 +129,26 @@ config BT_DRIVER_RX_HIGH_PRIO
 	int
 	default 6
 
+config BT_CONN_TX_WQ
+	bool "Use a separate workqueue for connection TX notify processing"
+	help
+	  Use a separate workqueue instead of system workqueue for
+	  bt_conn_tx_notify processing. The option can be used to make Bluetooth
+	  stack more independent from the system workqueue.
```

Comment:
There should be tests verifying that the stack works with and without this option enabled, right?

Should there also be tests that verifies that ATT cannot time out/proves it does when the system workqueue is blocked for a long time?

Reply:
I think we should just insist on having this enabled and remove the option. Doing so will remove the need for some hacks in the code like https://github.com/zephyrproject-rtos/zephyr/pull/79258/files#diff-00881f70fc7ec49b966fed79bd6a5a8ed5357599b241b5b029736bbc97683f83R275-R288

Reply:
Did we just remove one thread from BT and replace it with the system workqueue, and now adding a new thread to offload from the system workqueue?

Reply:
> I think we should just insist on having this enabled and remove the option. Doing so will remove the need for some hacks in the code like https://github.com/zephyrproject-rtos/zephyr/pull/79258/files#diff-00881f70fc7ec49b966fed79bd6a5a8ed5357599b241b5b029736bbc97683f83R275-R288

Removing the Kconfig option and making the feature enabled by default. I will also modify  https://github.com/zephyrproject-rtos/zephyr/pull/79258/files#diff-00881f70fc7ec49b966fed79bd6a5a8ed5357599b241b5b029736bbc97683f83R275-R288 (I noticed that you do not like this approach)

Reply:
> Did we just remove one thread from BT and replace it with the system workqueue, and now adding a new thread to offload from the system workqueue?

It seems that reusing system workqueue causes problems here. Because of that we may need to change approach here a bit.

Reply:
> Did we just remove one thread from BT and replace it with the system workqueue, and now adding a new thread to offload from the system workqueue?

We went from removing a thread, turning it into a non-blocking work-item and placing it on the system work queue, to then having people noticing that blocking the system work queue actually blocks the system work queue. But since "it worked before", it's our fault and the conclusion is that we cannot rely on the system work queue to not be blocked. We are now adding a second system work queue that is explicitly only for non-blocking work.

Reply:
> Right, but what we did gain from doing that then?

Architecturally, we removed a long-lived thread.

Ideally the new thread would not be in Bluetooth subsystem, but rather a service from the OS. There could be a kconfig "CONFIG_I_PROMISE_ONLY_SHORT_LIVED_TASKS_ON_SYSTEM_WORK_QUEUE". When selected the system work queue can be used.

> If we do need an internal workqueue for BT, then should we consider merging this with the BT_LONG_WQ? BT_LONG_WQ is not blocking, but may take a while to process, so not sure if we can use that

The long queue time will be a problem.

Reply:
> > Architecturally, we removed a long-lived thread and made it into a short-lived task.
> 
> Doesn't creating a new workqueue also create a long-living thread?
> 
> `k_work_queue_start` is basically just a wrapper around `k_thread_create` and `k_thread_start`, so we've just replaced a long-living thread with another long-living thread as far as I can tell.

I am worried that we may need to accept this approach (at least temporarily), because seemingly we cannot rely on the system work queue to not be blocked. Of course I agree that we could optimize the solution further later.

Reply:
Yes. The the number of threads as a metric will tell you that nothing has changed.

But the new thread does not need to be exclusive to the work that "BT TX" did before. It can be shared with any other short-lived tasks. Potentially system wide.

Reply:
> But the new thread does not need to be exclusive to the work that "BT TX" did before. It can be shared with any other short-lived tasks. Potentially system-wide.

"The new thread" here referring to system-wide thread this PR isn't adding, right? 

What I am concerned about is that this is a temp patch (which it sounds like it is), fixing a problem that isn't referenced nor described, effectively going back to the BT stack still using 3 threads, where the thread/workqueue added by this PR is also exclusively for TX (as it is only used for the TX notify). 

We are thus once again modifying the thread where the TX callbacks are called from, which is IMO not something we should take lightly.

Reply:
> We are thus once again modifying the thread where the TX callbacks are called from, which is IMO not something we should take lightly.

This is a valid point. I wasn't aware we are doing TX callbacks from tx_work. It looks like we are if the connection is disconnected or a bt_send error occurs. Thanks for the reminder. We should get rid of that somehow and consolidate the callback invocations on different thread. I maintain that blocking callbacks is stealing, and there is no way to fix this without green threads like in Go, but we can make the system slightly more resilient by having a "sacrificial" thread.

But switching over to a separate thread again will not make anything worse. Except, we might need to ensure the stack size is the same.

Reply:
> But switching over to a separate thread again will not make anything worse. Except, we might need to ensure the stack size is the same.

No, but what does it solve? 
The workqueue/thread added by this PR still calls the TX callback in `tx_notify_process`, so an application can still block that workqueue, so aren't we just back to the same? Sure, we reduce the number of places where the thread can be blocked by higher layers (as it's only a single callback), but it doesn't entirely prevent it.

Reply:
Did you decide upon that you didn't want to add any form of test coverage as part of this PR?

Reply:
@rugeGerritsen I think we should definitely have some tests in BSIM using this.

Reply:
If needed, maybe I could open a separate PR that enables the feature by default to run all of the CI tests with the feature enabled (for additional validation of the feature)?

Reply:
I would argue that we should run at least _some_ tests in this PR, otherwise we are adding a untested Kconfig. It makes more sense to test before you merge a feature, rather than after :D

Reply:
Let's enable feature by default on a separate test PR to run CI tests on it then: https://github.com/zephyrproject-rtos/zephyr/pull/79713

Reply:
> Let's enable feature by default on a separate test PR to run CI tests on it then: #79713

Can you explain why you are not adding tests in this PR? We need the tests regardless